lib/src/validator/leak_detection.dart

```diff
@@ -0,0 +1,799 @@
+import 'package:source_span/source_span.dart';
+
+import '../entrypoint.dart';
+import '../ignore.dart';
+import '../validator.dart';
+
+/// [Utf8Codec] which allows malformed strings.
+const _utf8AllowMalformed = Utf8Codec(allowMalformed: true);
+
+/// Link to the documentation for the `false_secrets` key in `pubspec.yaml`.
+const _falseSecretsDocumentationLink = 'https://dart.dev/go/false-secrets';
+
+/// A validator that validates attempts to find secrets that are about to be
+/// accidentally leaked.
+@sealed
+class LeakDetectionValidator extends Validator {
+  LeakDetectionValidator(Entrypoint entrypoint) : super(entrypoint);
+
+  @override
+  Future<void> validate() async {
+    // Load `false_secrets` from `pubspec.yaml`.
+    final falseSecrets = Ignore(
+      entrypoint.root.pubspec.falseSecrets,
+      ignoreCase: Platform.isWindows || Platform.isMacOS,
+    );
+
+    final pool = Pool(20); // don't read more than 20 files concurrently!
+    final leaks = await Future.wait(entrypoint.root.listFiles().map((f) async {
+      final relPath = entrypoint.root.relative(f);
+
+      // Skip files matching patterns in `false_secrets`
+      final nixPath = p.posix.joinAll(p.split(relPath));
+      if (falseSecrets.ignores(nixPath)) {
+        return <LeakMatch>[];
+      }
+
+      if (_filesToIgnore.any((p) => p.allMatches(relPath).isNotEmpty)) {
+        return <LeakMatch>[];
+      }
+
+      String text;
+      try {
+        // On Windows, we can't open some files without normalizing them
+        final file = File(p.normalize(p.absolute(f)));
+        text = await pool.withResource(
+          () async => await file.readAsString(encoding: _utf8AllowMalformed),
+        );
+      } on IOException {
+        // Pass, ignore files we can't read, let something else error later!
+        return <LeakMatch>[];
+      }
+
+      return leakPatterns
+          .map((p) => p.findPossibleLeaks(relPath, text))
+          .expand((i) => i);
+    })).then((lists) => lists.expand((i) => i).toList());
+
+    // Convert detect leaks to errors, if we have more than 3 then we print
+    // the first 2 leaks, followed by a general summary of leaks.
+    //
+    // This way we never return more than 3 errors, and we always show a 2-3
+    // samples leaks that very concretely demonstrates the strings we're
+    // worried about leaking.
+    if (leaks.length > 3) {
+      errors.addAll(leaks.take(2).map((leak) => leak.toString()));
+
+      final files = leaks
+          .map((leak) => leak.span.sourceUrl.toFilePath(windows: false))
+          .toSet()
+          .toList(growable: false)
+        ..sort();
+      final s = files.length > 1 ? 's' : '';
+
+      errors.add([
+        '${leaks.length} potential leak detected in ${files.length} file$s:',
+        ...files.take(10).map((f) => '- /$f'),
+        if (files.length > 10) '...',
+        '',
+        'Add a git-ignore style patterns to `false_secrets` in `pubspec.yaml`',
```

Comment:
```suggestion
        'Add a git-ignore style pattern to `false_secrets` in `pubspec.yaml`',
```

Reply:
Keeping it plural, but changing `Add a` to `Add`